add data migration info (#251)
cossacklabs/product-docs: content/acra/guides/integrating-acra-server-into-infrastructure/data_migration.md

# Data migration

How to prepare a database to work with AcraServer.

## Data migration strategies in general

Any data migration strategies have two main steps: (1) migrate data schema and (2) encrypt existing data.

> Schema migration includes changing column's types to `binary` (`bytea`) for using data encryption via Acra (not required for tokenization).

Reviewer: ```suggestion
Schema migration includes changing column's types to `binary` (`bytea`/`blob`) for using data encryption via Acra (not required for tokenization).
```

There are several approaches for **schema migration**:

- A. Stop application and producing data, change column's types with ALTER TABLE, resume the application.
- B. Create neighbour column “neighbour_column_name" with required column type, migrate here all existing data with type converting, lock database and switch columns with each other. Rename “column_name" to “_column_name" and “neighbour_column_name" to “column_name".

> - C. Create new database with required schema, migrate parallel current data with type converting, lock the current database and migrate the remaining data. Switch to a new database.

Author: @G1gg1L3s it would really help if you make suggestions instead of comments :)

> - C. Create new database with required schema, migrate parallel current data with type converting, lock the current database and migrate the remaining data. Switch to a new database.

Author: ```suggestion
- C. Create new database with required schema, migrate current data with type converting, lock the current database and migrate the remaining data. Switch to a new database.
```

**Encrypting the existing data**:

> - A. Write script that will query data from database and then pass them through Acra via UPDATE command that save same data. Acra will change plaintext in UPDATE command with ciphertext transparently. We have dummy scripts, they require populating with real schema.

Reviewer 2: > That save same data

Is somehow confusing, I would remove this phrase:

```suggestion
- A. Write script that will query data from database and then pass them through Acra via UPDATE command. Acra will change plaintext in UPDATE command with ciphertext transparently. We have dummy scripts, they require populating with real schema.
```
Also,
>   We have dummy scripts, they require populating with real schema.

I don't get: what scripts? :confused:

> - B. Write script that query data from database, encrypt it using AcraTranslator Bulk API and save encrypted data to database. It will require temporary launching AcraTranslator instance, and destroying it after migration.

Reviewer 2: ```suggestion
- B. Write a script that queries data from the database, encrypts it using AcraTranslator Bulk API and saves encrypted data to the database. It will require temporary launching AcraTranslator instance, and destroying it after migration.
```

> - C. [Acra EE only](/acra/enterprise-edition/) Use [acra-rotate utility](/acra/configuring-maintaining/general-configuration/acra-rotate/), write a script that iterates over all tables that need encryption, and pass `SELECT` query that fetch plaintext and `UPDATE` query that change same data in the table with encrypted. AcraRotate will query plaintext data, encrypt it and use `UPDATE` query to save it in database.

Reviewer 2: ```suggestion
- C. [Acra EE only](/acra/enterprise-edition/) Use [acra-rotate utility](/acra/configuring-maintaining/general-configuration/acra-rotate/), write a script that iterates over all tables that need encryption, and passes `SELECT` query that fetches plaintext and `UPDATE` query that replaces the same data in the table with encrypted. AcraRotate will query plaintext data, encrypt it and use `UPDATE` query to save it in database.
```
